### ISSUE TITLE
Bad user experience trying bare-bones web hello-world

### ISSUE THREAD

**Issue by johnniwinther** [dart-lang/sdk#43061]
I've had a very bad user experience trying out bare-bones web hello-world.

System:
IntelliJ IDEA 2020.2 (Community Edition) on Windows 7 6.1
Dart plugin version 202.6397.47
Dart SDK 2.9.1

Steps:

1) Create New Project -> Dart -> Generate sample content -> Bare-bones Web App

pubspec.yaml:
```
name: helloworld
description: An absolute bare-bones web app.
# version: 1.0.0
#homepage: https://www.example.com
#author: Johnni <email@example.com>

environment:
  sdk: '>=2.2.0 <3.0.0'

#dependencies:
#  path: ^1.4.1

dev_dependencies:
  build_runner: ^1.1.2
  build_web_compilers: ^1.0.0
  pedantic: ^1.0.0
```

2) Right-click on web/index.html -> Run 'web/index.html'

```
C:\Users\Johnni\dart-sdk\bin\pub.bat global run webdev daemon web:53322 --launch-app=web/index.html
webdev could not run for this project.
The `build_web_compilers` version – 1.2.3 – is not within the allowed constraint – >=2.6.1 <3.0.0.

Process finished with exit code 78
```

3) Surprising message. `build_web_compilers` has version `^1.0.0` according to `pubspec.yaml` so why it complains about `1.2.3` I have no idea. Also where does the constraint `>=2.6.1 <3.0.0` come from? The closest to it is the `>=2.2.0 <3.0.0` constraint on `sdk`. Since there is no help at this point, I click 'Pub get' in the 'pubspec.yaml' editor.

```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat get
Resolving dependencies...
Got dependencies!
Process finished with exit code 0
```

4) Looks good. Running 'web/index.html':

```
C:\Users\Johnni\dart-sdk\bin\pub.bat global run webdev daemon web:53322 --launch-app=web/index.html
webdev could not run for this project.
The `build_web_compilers` version – 1.2.3 – is not within the allowed constraint – >=2.6.1 <3.0.0.

Process finished with exit code 78
```

5) Ok, no change apparently. Try 'Pub upgrade' because the button is there (I don't know what it does as opposed to `pub get`):
```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat upgrade
Resolving dependencies...
  analyzer 0.35.4 (0.39.17 available)
  archive 2.0.13
  args 1.6.0
  async 2.4.2
  bazel_worker 0.1.25
  build 1.2.2 (1.3.0 available)
  build_config 0.4.2
  build_daemon 2.1.4
  build_modules 1.0.11 (2.10.1 available)
  build_resolvers 1.2.1 (1.3.11 available)
  build_runner 1.9.0 (1.10.2 available)
  build_runner_core 5.1.0 (6.0.1 available)
  build_web_compilers 1.2.3 (2.11.0 available)
  built_collection 4.3.2
  built_value 7.1.0
  charcode 1.1.3
  checked_yaml 1.0.2
  code_builder 3.4.0
  collection 1.14.13
  convert 2.1.1
  crypto 2.1.5
  dart_style 1.2.4 (1.3.6 available)
  fixnum 0.10.11
  front_end 0.1.14 (0.1.29 available)
  glob 1.2.0
  graphs 0.2.0
  http 0.12.2
  http_multi_server 2.2.0
  http_parser 3.1.4
  io 0.3.4
  js 0.6.2
  json_annotation 2.4.0 (3.0.1 available)
  kernel 0.3.14 (0.3.29 available)
  logging 0.11.4
  matcher 0.12.9
  meta 1.2.2
  mime 0.9.7
  node_interop 1.1.1
  node_io 1.1.1
  package_config 1.9.3
  package_resolver 1.0.10
  path 1.7.0
  pedantic 1.9.2
  pool 1.4.0
  protobuf 1.0.1
  pub_semver 1.4.4
  pubspec_parse 0.1.5
  quiver 2.1.3
  scratch_space 0.0.4+2
  shelf 0.7.9
  shelf_web_socket 0.2.3
  source_maps 0.10.9
  source_span 1.7.0
  stack_trace 1.9.5
  stream_channel 2.0.0
  stream_transform 1.2.0
  string_scanner 1.0.5
  term_glyph 1.1.0
  timing 0.1.1+2
  typed_data 1.2.0
  watcher 0.9.7+15
  web_socket_channel 1.1.0
  yaml 2.2.1
No dependencies changed.
11 packages have newer versions incompatible with dependency constraints.
Try `pub outdated` for more information.
Process finished with exit code 0
```

6) No changes, and running `web/index.html` still doesn't work. Since it tells me to try `pub outdated` I do that:

```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat outdated
Dependencies         Current  Upgradable  Resolvable  Latest    
all up-to-date      

dev_dependencies    
build_runner         *1.9.0   *1.9.0      *1.10.0     1.10.2    
build_web_compilers  *1.2.3   *1.2.3      2.11.0      2.11.0    

transitive dependencies: all up-to-date

transitive dev_dependencies
_fe_analyzer_shared  -        -           7.0.0       7.0.0    
analyzer             *0.35.4  *0.35.4     0.39.17     0.39.17  
build                *1.2.2   *1.2.2      1.3.0       1.3.0    
build_modules        *1.0.11  *1.0.11     2.10.1      2.10.1    
build_resolvers      *1.2.1   *1.2.1      1.3.11      1.3.11    
build_runner_core    *5.1.0   *5.1.0      *5.2.0      6.0.1    
cli_util             -        -           0.2.0       0.2.0    
csslib               -        -           0.16.2      0.16.2    
dart_style           *1.2.4   *1.2.4      1.3.6       1.3.6    
front_end            *0.1.14  *0.1.14     -           0.1.29    
html                 -        -           0.14.0+3    0.14.0+3  
json_annotation      *2.4.0   *2.4.0      3.0.1       3.0.1    
kernel               *0.3.14  *0.3.14     -           0.3.29    

9  dependencies are constrained to versions that are older than a resolvable version.
To update these dependencies, edit pubspec.yaml.
Process finished with exit code 0
```

7) It tells me to update the dependencies but I have no clue what the printed information is or how it relates to the
pubspec.yaml file. I have no idea what the columns `Current`, `Upgradable`, `Resolvable` and `Latest` mean. Since I have `build_runner` and `build_web_compilers` in my pubspec.yaml I try using the values from each column, one at a time.

8) I set the version numbers of `build_runner` and `build_web_compilers` to the `Current`/`Upgradable` values `*1.9.0` and `*1.2.3`, respectively. Not knowing what to do next I try running both `pub outdated` and `pub upgrade`, both give the result:

```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat upgrade
Error on line 14, column 17 of pubspec.yaml: Undefined alias.
   â•·
14 â”‚   build_runner: *1.9.0
   â”‚                 ^^^^^^
   â•µ
Process finished with exit code 65
```

9) So apparently `*` wasn't part of the version number. Why it's part of the print-out then, baffles me. Also note the apparent use of non-ascii in the error message giving rise to the `â•·`, `â”‚`, `â”‚`, and `â•µ`.

10) Ok, I remove the `*` from the version numbers and try `pub upgrade` again.

```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat upgrade
Resolving dependencies...
  analyzer 0.35.4 (0.39.17 available)
  archive 2.0.13
  args 1.6.0
  async 2.4.2
  bazel_worker 0.1.25
  build 1.2.2 (1.3.0 available)
  build_config 0.4.2
  build_daemon 2.1.4
  build_modules 1.0.11 (2.10.1 available)
  build_resolvers 1.2.1 (1.3.11 available)
  build_runner 1.9.0 (1.10.2 available)
  build_runner_core 5.1.0 (6.0.1 available)
  build_web_compilers 1.2.3 (2.11.0 available)
  built_collection 4.3.2
  built_value 7.1.0
  charcode 1.1.3
  checked_yaml 1.0.2
  code_builder 3.4.0
  collection 1.14.13
  convert 2.1.1
  crypto 2.1.5
  dart_style 1.2.4 (1.3.6 available)
  fixnum 0.10.11
  front_end 0.1.14 (0.1.29 available)
  glob 1.2.0
  graphs 0.2.0
  http 0.12.2
  http_multi_server 2.2.0
  http_parser 3.1.4
  io 0.3.4
  js 0.6.2
  json_annotation 2.4.0 (3.0.1 available)
  kernel 0.3.14 (0.3.29 available)
  logging 0.11.4
  matcher 0.12.9
  meta 1.2.2
  mime 0.9.7
  node_interop 1.1.1
  node_io 1.1.1
  package_config 1.9.3
  package_resolver 1.0.10
  path 1.7.0
  pedantic 1.9.2
  pool 1.4.0
  protobuf 1.0.1
  pub_semver 1.4.4
  pubspec_parse 0.1.5
  quiver 2.1.3
  scratch_space 0.0.4+2
  shelf 0.7.9
  shelf_web_socket 0.2.3
  source_maps 0.10.9
  source_span 1.7.0
  stack_trace 1.9.5
  stream_channel 2.0.0
  stream_transform 1.2.0
  string_scanner 1.0.5
  term_glyph 1.1.0
  timing 0.1.1+2
  typed_data 1.2.0
  watcher 0.9.7+15
  web_socket_channel 1.1.0
  yaml 2.2.1
No dependencies changed.
11 packages have newer versions incompatible with dependency constraints.
Try `pub outdated` for more information.
Process finished with exit code 0
```

11) Apparently no success then, it still complains about incompatible versions. Let's try the `Resolvable` column, `*1.10.0` and `2.11.0`, without `*` this time. I run `pub upgrade`:

```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat upgrade
Resolving dependencies...
+ _fe_analyzer_shared 7.0.0
> analyzer 0.39.17 (was 0.35.4)
  archive 2.0.13
  args 1.6.0
  async 2.4.2
  bazel_worker 0.1.25
> build 1.3.0 (was 1.2.2)
  build_config 0.4.2
  build_daemon 2.1.4
> build_modules 2.10.1 (was 1.0.11)
> build_resolvers 1.3.11 (was 1.2.1)
> build_runner 1.10.0 (was 1.9.0) (1.10.2 available)
> build_runner_core 5.2.0 (was 5.1.0) (6.0.1 available)
> build_web_compilers 2.11.0 (was 1.2.3)
  built_collection 4.3.2
  built_value 7.1.0
  charcode 1.1.3
  checked_yaml 1.0.2
+ cli_util 0.2.0
  code_builder 3.4.0
  collection 1.14.13
  convert 2.1.1
  crypto 2.1.5
+ csslib 0.16.2
> dart_style 1.3.6 (was 1.2.4)
  fixnum 0.10.11
  glob 1.2.0
  graphs 0.2.0
+ html 0.14.0+3
  http_multi_server 2.2.0
  http_parser 3.1.4
  io 0.3.4
  js 0.6.2
> json_annotation 3.0.1 (was 2.4.0)
  logging 0.11.4
  matcher 0.12.9
  meta 1.2.2
  mime 0.9.7
  node_interop 1.1.1
  node_io 1.1.1
  package_config 1.9.3
  path 1.7.0
  pedantic 1.9.2
  pool 1.4.0
  protobuf 1.0.1
  pub_semver 1.4.4
  pubspec_parse 0.1.5
  quiver 2.1.3
  scratch_space 0.0.4+2
  shelf 0.7.9
  shelf_web_socket 0.2.3
  source_maps 0.10.9
  source_span 1.7.0
  stack_trace 1.9.5
  stream_channel 2.0.0
  stream_transform 1.2.0
  string_scanner 1.0.5
  term_glyph 1.1.0
  timing 0.1.1+2
  typed_data 1.2.0
  watcher 0.9.7+15
  web_socket_channel 1.1.0
  yaml 2.2.1
These packages are no longer being depended on:
- front_end 0.1.14
- http 0.12.2
- kernel 0.3.14
- package_resolver 1.0.10
Changed 17 dependencies!
2 packages have newer versions incompatible with dependency constraints.
Try `pub outdated` for more information.
Process finished with exit code 0
```

12) Still no success, I guess, it still talks about some incompatible versions and tells me to try `pub outdated`. Trying the `Latest` column, `1.10.2` and `2.11.0`, and run `pub upgrade`:
```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat upgrade
Resolving dependencies...
The current Dart SDK version is 2.9.1.

Because pub depends on build_runner >=1.10.2 which requires SDK version >=2.10.0-0.0 <3.0.0, version solving failed.
Process finished with exit code 1
```

13) Still no success.

14) I remember `any` being a possible value for package versions, and since I really don't care about versions, I'm not using these packages - I just want to run the app, I try `any` for `build_runner`, `build_web_compilers`, and `pedantic` (it hasn't caused any problems yet, but you never know!).

```
Working dir: C:\Users\Johnni\workspace\pub
C:\Users\Johnni\dart-sdk\bin\pub.bat upgrade
Resolving dependencies...
  _fe_analyzer_shared 7.0.0
  analyzer 0.39.17
  archive 2.0.13
  args 1.6.0
  async 2.4.2
  bazel_worker 0.1.25
  build 1.3.0
  build_config 0.4.2
  build_daemon 2.1.4
  build_modules 2.10.1
  build_resolvers 1.3.11
  build_runner 1.10.0 (1.10.2 available)
  build_runner_core 5.2.0 (6.0.1 available)
  build_web_compilers 2.11.0
  built_collection 4.3.2
  built_value 7.1.0
  charcode 1.1.3
  checked_yaml 1.0.2
  cli_util 0.2.0
  code_builder 3.4.0
  collection 1.14.13
  convert 2.1.1
  crypto 2.1.5
  csslib 0.16.2
  dart_style 1.3.6
  fixnum 0.10.11
  glob 1.2.0
  graphs 0.2.0
  html 0.14.0+3
  http_multi_server 2.2.0
  http_parser 3.1.4
  io 0.3.4
  js 0.6.2
  json_annotation 3.0.1
  logging 0.11.4
  matcher 0.12.9
  meta 1.2.2
  mime 0.9.7
  node_interop 1.1.1
  node_io 1.1.1
  package_config 1.9.3
  path 1.7.0
  pedantic 1.9.2
  pool 1.4.0
  protobuf 1.0.1
  pub_semver 1.4.4
  pubspec_parse 0.1.5
  quiver 2.1.3
  scratch_space 0.0.4+2
  shelf 0.7.9
  shelf_web_socket 0.2.3
  source_maps 0.10.9
  source_span 1.7.0
  stack_trace 1.9.5
  stream_channel 2.0.0
  stream_transform 1.2.0
  string_scanner 1.0.5
  term_glyph 1.1.0
  timing 0.1.1+2
  typed_data 1.2.0
  watcher 0.9.7+15
  web_socket_channel 1.1.0
  yaml 2.2.1
No dependencies changed.
2 packages have newer versions incompatible with dependency constraints.
Try `pub outdated` for more information.
Process finished with exit code 0
```

15) Apparently, this doesn't help. Having given up at this point I try running 'web/index.html' a last time. And now it works - I have no clue why.

In conclusion, the user experience at all the steps of this journey has been horrible. If this was my first encounter with Dart it would probably be the last.

Post-script: Going back and retrying the various version numbers it turns out that `1.10.0`/`2.10.0` works. Apparently you have to try run `web/index.html` to test if it works and the output of `pub upgrade` is no indication of success - even though it can be an indication of failure.

**Comment by devoncarew** [comment]
@kevmoo @mit-mit 

Johnni, I think the issue is that `pub get` succeeds on the project, but that `webdev serve` fails (w/ some strange version constraint issue). I think your point is that the difference between the two isn't meaningful for new users - they won't know how to resolve the issue.

Kevin, Michael - I'm not sure if this is a specific issue we need to address, or a larger problem. I agree w/ Johnni that this is a super rough getting started experience, for something that should just work. We are shipping the 'web-simple' stagehand template as part of dartdev, so this is an issue that would be exposed to all dart developers.

Here's the error:

```
The `build_web_compilers` version – 1.2.3 – is not within the allowed constraint – >=2.6.1 <3.0.0.
```

Concretely:
- I think we need adjust the build_web_compilers constraint in stagehand: https://github.com/dart-lang/stagehand/blob/master/templates/web-simple/pubspec.yaml
- this error message should be improved to make it clearer and more actionable
- we should have some automated CI which will warn us when version solves will fail (run nightly?)
- once a fix is in stagehand, we'll need to publish that and bring it into the dart sdk (so dartdev can consume it; cc @jwren)

Johnni, can you open an issue against stagehand for the first item?

**Comment by johnniwinther** [comment]
The constraints in https://github.com/dart-lang/stagehand/blob/master/templates/web-simple/pubspec.yaml don't match the ones I had, so I'm not sure that they need updating. Seems like the problems is that I somehow had a stale template.

**Comment by devoncarew** [comment]
OK, closing this issue then. You might consider filing an issue on stagehand - we should periodically (via the CI?) validate that we can create new projects using the current set of package constraints.
